<!-- Desktop 
<link rel="stylesheet" href="../bower_components/jquery-ui/themes/smoothness/jquery-ui.min.css">
<script src="../bower_components/jquery-ui/ui/core.js"></script>
<script src="../bower_components/jquery-ui/ui/widget.js"></script>
<script src="../bower_components/jquery-ui/ui/accordion.js"></script>
-->
<div class="science-app" data-app-name="araport-geneslider">
       <!-- ///////////////////////////////////////////////////////////////////// -->
        <!-- Load a Gene Section -->
        <div id="araport-geneslider-loadDataBox" class="text-center">
            <div class="col-lg-12">
                <!-- Enter AGI ID box -->
                <h2 id="araport-geneslider-title">Select an <i>Arabidopsis</i> gene to view conserved regions across species:</h2>
                <div class="row" style="margin-bottom:20px;">
                    <!-- left col -->
                    <div class="col-md-2"></div>

                    <!-- center col -->
                    <div class="col-md-8">
                        <div class="row">

                              <!-- Before slider -->
                              <div class="col-sm-4">
                                  <div class="row">
                                        <div class="col-sm-12 text-center">
                                            <label id="araport-geneslider-beforelabel">Before (0)</label>
                                        </div>
                                  </div>

                                  <div class="row">
                                        <div class="col-sm-2" style="text-align:left;">1000</div>
                                        <div class="col-sm-8" style="padding-right:0px;">
                                            <input type="range" class="rangesliders" id="araport-geneslider-beforeSlider" name="before" value="0" min="-1000" max="0" step="10">
                                        </div>
                                        <div class="col-sm-2" style="text-align:right;">0</div>
                                  </div>                                
                              </div>

                              <!-- Locus Text box -->
                              <div class="col-sm-4 text-center">
                                  <label class="control-label">Locus</label>
                                  <div class="row" style="display: flex;">
                                      <input class="form-control" type="text" id="araport-geneslider-user_agi" placeholder="Enter an AGI ID here" >

                                      <!-- Go button -->
                                      <span class="btn btn-success btn-file" id="araport-geneslider-go" style="margin-left: 5px;">Go</span>
                                  </div>
                              </div>

                              <!-- After slider -->
                              <div class="col-sm-4">
                                  <div class="row">
                                        <div class="col-sm-12 text-center">
                                            <label id="araport-geneslider-afterlabel">After (0)</label>
                                        </div>
                                  </div>
                                  <div class="row">
                                        <div class="col-sm-2" style="text-align:right;">0</div>
                                        <div class="col-sm-8" style="padding-right:0px;">
                                            <input type="range" class="rangesliders" id="araport-geneslider-afterSlider" name="after" value="0" min="0" max="1000" step="10">
                                        </div>             
                                         <div class="col-sm-2" style="text-align:left;">1000</div>
                                  </div>
                              </div>

	                      </div>
				            </div>					
                    <!-- right col -->
                    <div class="col-md-2"></div>
				          </div> 
                </div>   

                <!-- upload file & aboout buttons -->
                <div>
                    <span class="btn btn-default btn-sm" id="araport-geneslider-loadUpstream">+1000 Bases Upstream</span>
                    <span class="btn btn-default btn-sm" id="araport-geneslider-about" data-toggle="modal" data-target="#aboutModal">About</span>
                    <span class="btn btn-default btn-sm" id="araport-geneslider-contact" data-toggle="modal" data-target="#contactModal">Contact</span>
                    <span class="btn btn-default btn-sm" id="araport-geneslider-file" >Upload Your Own FASTA File</span>
					<input type="file" id="araport-geneslider-fileUpload" style="display: none;" >
                    <span class="btn btn-default btn-sm" id="araport-geneslider-loadDownstream">+1000 Bases Downstream</span>
                </div>
          </div>


          <!-- ///////////////////////////////////////////////////////////////////// -->
          <!-- Visualization Window -->
  				<div class="row">
  					<div class="col-md-8">
  						<!-- This is for Desktop version 
  						<canvas id="araport-geneslider-canvas" data-processing-sources="../bower_components/GeneSlider_Bower_Package/GeneSlider.pde"></canvas>	
						-->
  						<!-- This is for Araport Live version --> 
  						<canvas id="araport-geneslider-canvas"></canvas>							
  					</div>
  				</div>


          <!-- ///////////////////////////////////////////////////////////////////// -->
          <!-- About Modal -->
          <div class="modal fade" id="aboutModal" tabindex="-1" role="dialog" aria-labelledby="About button" aria-hidden="true">
              <div class="modal-dialog">
                  <div class="modal-content">
                      <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal" aria-hidden="true">x</button>
                          <h1><a href="http://bar.utoronto.ca"><img src="http://bar.utoronto.ca/~jwaese/GeneSlider/images/barLogo.png" style="margin-right:10px" alt="BAR logo"></a>GeneSlider</h1>
                      </div>

                      <div class="modal-body" id="aboutBody">
                          <p>Sequence logos are a well-known bioinformatic data visualization paradigm for displaying conservation and entropy in aligned sequences (Schneider and Stephens, 1990). Each of the residues in a sequence are represented by a series of stacked characters in which the height the stack signifies the information content, reflecting the degree of conservation, for that position, and the height of the characters signifies the frequency of that residue. In the case of Gene Slider running on Araport.org, we have pre-processed 90,000+ conserved non-coding regions across the Brassicaceae (Haudry et al., 2013), indexed to the TAIR10 Col-0 Arabidopsis thaliana sequence, for easy exploration. Further, we have mapped transcription factor binding sites from the JASPAR database (Mathelier et al., 2014) and from Weirauch et al. 2014 to the TAIR10 genome sequence to permit the identification of regions that are both conserved and may contain transcription factor binding sites (see forthcoming Waese et al. publication for methodology).</p>

                          <p>Traditionally, sequence logos have been generated with tools such as WebLogo (Crooks et al., 2004) and Seq2Logo (Thomsen and Nielsen, 2012) in which any sequence beyond 30 or so bases must be broken into separate rows. This makes it difficult to identify motifs that span multiple rows because the visual disconnect of a line break interferes with our ability to recognize elements based on preattentive visual features such as size, density, and number (Healey and Enns, 2012). A new web tool called Skylign (Wheeler et al., 2014) addresses this problem by generating long sequence logos that can be zoomed in and out of, however it is still difficult to identify amino acid motifs when there is “wobble” in one or more of the positions because it is inherently difficult to distinguish targets of interest from a set of distracters (Quinlan and Humphreys, 1987).</p>

                          <p>Gene Slider addresses these challenges by combining the ability to make long sequence logos with a search panel that automatically highlights motifs of interest, even when there is a high degree of wobble. Slider controls for each position of a search query may be used to assign thresholds to specify whether the search function will recognize partially conserved matches. A navigation slider with a “thumbnail” background at the bottom of the screen indicates which region of the sequence is currently being displayed. Coloured markers along this slider indicate where query motifs have been identified, making it easy to see where multiple matches have been found. Figure 1 shows a portion of an alignment in which there is only one occurrence of the motif AAACA. Without the search function it would be challenging to find this motif because of the entropy in positions 3 and 4 of the match.</p>

                          <figure style="margin-bottom:20px;">
                          <img src="http://bar.utoronto.ca/~jwaese/GeneSlider/images/Figure1.png" style="width:100%;">
                          <figcaption>Fig. 1. The motif AAACA is highlighted according to the query in the search panel. The grey horizontal line superimposed over each column represents the number of sequences with a residue at that position, read from the scale on the right. The left scale denote the bit score (Schneider and Stephens, 1990) of the column, reflecting the degree of conservation.</figcaption>
                          </figure>

                          <p>Gene Slider loads FASTA formatted aligned sequences into a large matrix and then counts the frequency of each base per column. Bit scores for each base are calculated according to Schneider & Stephens’s (1990) application of the Shannon entropy, H, formula, where:</p>
                          <p style="text-align:center;"><img src="http://bar.utoronto.ca/~jwaese/GeneSlider/images/shannons-formula-small.jpg" style="width:40%;">
                          </p>

                          <p>and p(x) is the frequency of each of the 20 possible amino acids or 4 possible nucleotides. In a given column, letter symbols are stacked in order of most abundant to least abundant and are scaled according to their frequency.</p>
                          
                          <p>This depiction of a sequence logo can be misleading because the original Schneiderman and Stephens’ calculation of column heights does not take insertion and deletion events into account. A residue that is only found in one sequence with gap characters in the other sequences will have the same bit score and column height as a residue that is conserved across every sequence. Gene Slider provides two methods for unmasking this information. First, we have drawn a horizontal grey line across each residue to indicate how many sequences are contributing to that residue’s bit score. An example of this can be seen in Figure 1 where positions 534-539 only have one species with a residue in that column. An alternate method for displaying this information involves manipulating the saturation level of each column’s color. This is demonstrated in Figure 2 (a zoomed out view of Figure 1) where a large insertion point upstream of the highlighted motif and two small insertions downstream are quickly identifiable thanks to our preattentive ability to quickly recognize subtle changes in luminance (Healey and Enns, 2012).</p>

                          <figure style="margin-bottom:20px;">
                          <img src="http://bar.utoronto.ca/~jwaese/GeneSlider/images/Figure2.png" style="width:100%;">
                          <figcaption>Fig. 2. The motif AAACA is highlighted after creating a sequence logo description of the query using the visual search panel.</figcaption>
                          </figure>

                          <p>Another solution to this problem is to create a new “weighted bit score” that factors the number of sequences contributing to a position’s bit score into the calculation of that position’s height. This is different from the Kullback-Leibler, Weighted Kullback-Leibler, Probability Weighted Kullback-Leibler and PSSMLogos that can be generated with Seq2Logo (Thomsen and Nielsen, 2012) to display amino acid depletion with various methods on the negative y-axis. The weighted bit score is calculated by simply multiplying the bit score by the number of residues in that column divided by the total number of sequences in the alignment. With this modification, highly conserved regions are not conflated with ones having just one or two conserved residues in an otherwise gapped region.</p>

                          <p>Gene Slider was written in Processing.js and runs in most web browsers. It can display a large number of sequences of any length, however performance decreases as the sequence length and sequence number goes up. Contemporary laptops can display sequences up to 3000 residues long without lagging. Data can be loaded directly from the users’ hard drive, however CNS data can be loaded by using an ADAMA adaptor which GETS data from a RESTful webservice running on the BAR (Toufighi et al., 2005).</p>

                          <h2>References</h2>

                          <p style="padding-left:30px; text-indent:-30px;">Crooks, G.E., Hon, G., Chandonia, J.-M., and Brenner, S.E. (2004). WebLogo: a sequence logo generator. Genome Res. 14, 1188–1190.
                          Haudry, A., Platts, A.E., Vello, E., Hoen, D.R., Leclercq, M., Williamson, R.J., Forczek, E., Joly-Lopez, Z., Steffen, J.G., Hazzouri, K.M., et al. (2013). An atlas of over 90,000 conserved noncoding sequences provides insight into crucifer regulatory regions. Nat. Genet. 45, 891–898.
                          Healey, C.G., and Enns, J.T. (2012). Attention and visual memory in visualization and computer graphics. IEEE Trans. Vis. Comput. Graph. 18, 1170–1188.</p>
                          
                          <p style="padding-left:30px; text-indent:-30px;"> Mathelier, A., Zhao, X., Zhang, A.W., Parcy, F., Worsley-Hunt, R., Arenillas, D.J., Buchman, S., Chen, C., Chou, A., Ienasescu, H., et al. (2014). JASPAR 2014: an extensively expanded and updated open-access database of transcription factor binding profiles. Nucleic Acids Res. 42, D142–D147.</p>

                         
                          <p style="padding-left:30px; text-indent:-30px;">Quinlan, P., and Humphreys, G. (1987). Visual Search for Targets Defined by Combinations of Color, Shape, and Size: An Examination of Task Constraints on Feature and Conjunction Searches. Percept. Psychophys. 41, 455–472.</p>
                          
                          <p style="padding-left:30px; text-indent:-30px;">Schneider, T.D., and Stephens, R.M. (1990). Sequence logos: a new way to display consensus sequences. Nucleic Acids Res. 18, 6097–6100.
                          Thomsen, M.C.F., and Nielsen, M. (2012). Seq2Logo: a method for construction and visualization of amino acid binding motifs and sequence profiles including sequence weighting, pseudo counts and two-sided representation of amino acid enrichment and depletion. Nucleic Acids Res. 40, W281–W287.</p>
                          
                          <p style="padding-left:30px; text-indent:-30px;">Wheeler, T.J., Clements, J., and Finn, R.D. (2014). Skylign: a tool for creating informative, interactive logos representing sequence alignments and profile hidden Markov models. BMC Bioinformatics 15, 7.</p>
                          
                          <p style="padding-left:30px; text-indent:-30px;"></p>
                          <p style="padding-left:30px; text-indent:-30px;"></p>
                          <p style="padding-left:30px; text-indent:-30px;"></p>


                      </div>        
                  </div>
              </div>
          </div>


          <!-- ///////////////////////////////////////////////////////////////////// -->
          <!-- Contact Modal -->
          <div class="modal fade" id="contactModal" tabindex="-1" role="dialog" aria-labelledby="Contact button" aria-hidden="true">
              <div class="modal-dialog">
                  <div class="modal-content">
                      <div class="modal-header">
                      <button type="button" class="close" data-dismiss="modal" aria-hidden="true">x</button>
                      <h2 class="modal-title" id="myModalLabel">Contact</h2>
                      </div>
                      <div class="modal-body">
                          <p>Gene Slider was written by Jamie Waese, Asher Pasha, David Guttman and Nicholas Provart at the Department of Cell and Systems Biology, University of Toronto.</p>
                          
                          <p>For more information, please contact: <a href="mailto:jamie.waese@mail.utoronto.ca">jamie.waese@mail.utoronto.ca</a></p>
                      </div>        
                  </div>
              </div>
          </div>



</div>
